coreos-koji-tagger/coreos_koji_tagger.py

```diff
@@ -264,69 +288,111 @@ def __call__(self, message: fedora_messaging.api.Message):
-                    if self.keytab_file:
-                        tag_builds(tag=tag, builds=buildstotagforthisrelease)
+            # Get a set of tuples of build name to tag to tag into
+            tuples = [(self.intermediate_tag.format(
+                        releasever=buildsinfo[x].releasever),
+                        buildsinfo[x].nvr)
+                        for x in buildstotag]
+            logger.info('Tagging the following (tag, nvr): \n\t%s' %
+                                            '\n\t'.join(map(str, tuples)))
+            if self.keytab_file:
+                with self.koji_client.multicall(strict=True) as m:
+                    for (tag, nvr) in tuples:
+                        m.tagBuild(tag=tag, build=nvr)
             logger.info('Tagging done')
 
+
+    def get_buildsinfo_from_rpmnevras(self, rpmnevras: set) -> dict:
+        # Given a set of rpm NEVRAs get a set of corresponding koji buildids
+        if not rpmnevras:
+            raise ValueError("No nevras to get_builds_from_rpmnevras")
+
+        buildsinfo = {} # dict of buildid -> BuildInfo object
+
+        # Grab info about each of the rpms
+        with self.koji_client.multicall(strict=True) as m:
+            nvras = [get_NVRA_from_NEVRA(nevra) for nevra in rpmnevras]
+            rpminfos = [m.getRPM(nvra, strict=True) for nvra in nvras]
+
+        for rpm in [r.result for r in rpminfos]:
+            buildsinfo[rpm['build_id']] = \
+                BuildInfo(id = rpm['build_id'],
+                          buildrootid = rpm['buildroot_id'])
+
+        # Grab info about each related build
+        # Grab info about each builds buildroot
+        buildinfo = {}
+        brinfo = {}
+        with self.koji_client.multicall(strict=True) as m:
+            for buildid in buildsinfo.keys():
+                buildinfo[buildid] = m.getBuild(buildid, strict=True)
+                brinfo[buildid] = \
+                    m.getBuildroot(buildsinfo[buildid].buildrootid, strict=True)
+
+        # Update the BuildInfo objects with the information we now have
+        # Now translate the buildroot tag name into a releasever and
+        # add it to the BuildInfo objects
+        for buildid in buildsinfo.keys():
+            buildsinfo[buildid].name = buildinfo[buildid].result['name']
+            buildsinfo[buildid].nvr = buildinfo[buildid].result['nvr']
+            # Apply a heuristic to the buildroot tag name in order to
+            # derive the releasever for this build so that we can then
+            # know what intermediate tag to tag it into
+            buildsinfo[buildid].releasever = \
+                get_releasever_from_buildroottag(
+                    brinfo[buildid].result['tag_name'])
+
+        return buildsinfo
+
+    def get_pkglist_in_tag(self, tag: str) -> set:
+        # Given a tag, return the set of packages in its pkglist
+        pkgs = self.koji_client.listPackages(tagID=tag)
+        return set([pkg['package_name'] for pkg in pkgs])
```

Comment:
Does `koji_client` guarantee that `package_name` will exist? If not we should probably use a safe getter.

Reply:
yes.. If it wasn't there we'd want an exception